### PR TITLE
Fix crash in prepareToFixMemberNames

### PR DESCRIPTION
**Pull request by gacholio**
Work by @jdmpapin with a couple of fixes from me.

Fixes: #18076

### CODE REVIEW

**Comment by pshipton** [comment]
jenkins compile xlinux jdk21

**Comment by pshipton** [comment]
@bmarwell I created a jdk21 build with this fix for https://github.com/eclipse-openj9/openj9/issues/18076, if you want to give it a try.

https://openj9-artifactory.osuosl.org/artifactory/ci-openj9/Build_JDK21_x86-64_linux_Personal/51/OpenJ9-JDK21-x86-64_linux-20231005-110509.tar.gz

**Comment by bmarwell** [comment]
Thanks! Will try right away tomorrow. The error was reproducible on our Linux x86 pipeline (but not on AIX ppc64), so that build is exactly what I can use for testing.

**Review by dmitripivkine — APPROVED** [review]
Well, I never like an idea to use `classUnloadMutex` under Exclusive to stop compilation. And now we are going to  have an exception on the top of it. Also we assume `isClassUnloadMutexHeldForRedefinition` should not be changed during GC (I think it is reasonable assumption).
Ok, we can go with this PR if we don't want to change sync mechanism fundamentally.

**Comment by pshipton** [comment]
jenkins compile aix jdk21

**Comment by bmarwell** [comment]
I can confirm the fix works in my Jenkins pipeline. Thanks to everyone! 🙌🏻

**Review by tajila — APPROVED** [review]

**Comment by tajila** [comment]
jenkins test sanity alinux64 jdk21

**Comment by tajila** [comment]
any comments @amicic

**Comment by amicic** [comment]
looks good, I paid attention to when isClassUnloadMutexHeldForRedefinition was changing I can see that its while we have exclusive (directly or inderectly through safe point) - so no race with GC queuing the flag

**Review by amicic — APPROVED** [review]

File: runtime/oti/j9nonbuilder.h [hunk 1]
```diff
@@ -5773,6 +5773,16 @@ typedef struct J9JavaVM {
 #else
 	omrthread_monitor_t classUnloadMutex;
 #endif
+	/* When GC runs as part of class redefinition, it could unload classes and
+	 * therefore try to acquire the class unload mutex, which is already held
+	 * to prevent JIT compilation from running concurrently with redefinition.
+	 * Such an acquisition attempt could deadlock if it occurs on a different
+	 * thread, even though the acquisition would be morally recursive.
```
Comment (keithc-ca) [inline runtime/oti/j9nonbuilder.h:5780]: Perhaps "morally" should be "normally" here?

Comment (jdmpapin) [reply runtime/oti/j9nonbuilder.h:5780]: An acquisition that occurs on a different thread is technically never a recursive acquisition. However, in this case it's tantamount to one, and the fact that it's occurring on another thread is only a technicality. So I meant *morally* recursive as opposed to *technically* recursive

**Pull request merged**